### ISSUE TITLE
Incorrect diagnostic with box patterns

### ISSUE THREAD

**Issue by xldenis** [rust-analyzer/rust-analyzer#7711]
It seems like box patterns cause RA to think a struct field isn't present. For example:

<img width="778" alt="image" src="https://user-images.githubusercontent.com/451373/108268337-a7c45280-716c-11eb-9889-ae9e2daac2eb.png">

I've managed to reprodcue this error on the following example:

```
struct A {s: Box<u32>, o: bool }

fn x(a:A) {
  let A { box s, o } = a;
}
```

It seems like binders inside `box` patterns aren't considered.

**Comment by Veykril** [comment]
Seems like we incorrectly lower RecordPatterns for field box patterns here. 
https://github.com/rust-analyzer/rust-analyzer/blob/20a911f3cc2beb0409ab71cc1560648374745f7f/crates/hir_def/src/body/lower.rs#L908-L918
The `field_name` call fails for box patterns as they aren't special cased in
https://github.com/rust-analyzer/rust-analyzer/blob/20a911f3cc2beb0409ab71cc1560648374745f7f/crates/syntax/src/ast/node_ext.rs#L380-L389
like it is for `IdentPat`

**Comment by lnicola** [comment]
Can these be nested arbitrarily (`box box foo`)? I'm not even sure.

**Comment by Veykril** [comment]
Looks like they can: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f2d0fd84a01671aecd315eb8af2d953c

**Comment by Veykril** [comment]
They can be nested in general patterns, but not in a RecordPatField it seems
